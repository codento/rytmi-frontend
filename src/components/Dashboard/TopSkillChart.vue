<template>
  <chart-card :size="size">
    <template slot="header">
      <div class="h3">
        Top skill
      </div>
    </template>
    <b-col>
      <p class="lead px-2">
        Our top skill is
        <strong>{{ topSkillLevel.name }}</strong>. <br>
        There are <strong>{{ topSkillLevel.numberOfPeople }}</strong>
        who's skill level is 4 or 5 and the average is
        <strong>{{ topSkillLevel.currentKnowledgeAverage }}</strong>
      </p>
    </b-col>
  </chart-card>
</template>

<script>
import ChartCard from './ChartCard'

export default {
  components: {
    ChartCard
  },
  props: {
    size: {
      type: Number,
      default: 6
    },
    skillInfo: {
      type: Object,
      required: true
    }
  },
  computed: {
    topSkillLevel () {
      let mostProficientSkill = {
        name: '',
        numberOfPeople: 0,
        currentKnowledgeAverage: 0
      }
      const skillId = Object.keys(this.skillInfo).reduce((accumulate, current) => {
        if (this.skillInfo[current].proficiencies.filter(this.filterTop).length >
             this.skillInfo[accumulate].proficiencies.filter(this.filterTop).length) {
          return current
        }
        return accumulate
      })
      const topSkill = this.skillInfo[skillId]
      mostProficientSkill = { name: topSkill.name,
        numberOfPeople: topSkill.willingness.filter(this.filterTop).length,
        currentKnowledgeAverage: topSkill.proficiencyAverage }
      return mostProficientSkill
    }
  },
  methods: {
    filterTop (val) {
      return val === 4 || val === 5
    }
  }
}
</script>
