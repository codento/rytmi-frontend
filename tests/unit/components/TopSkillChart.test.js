import { shallowMount, createLocalVue } from '@vue/test-utils'
import BootstrapVue from 'bootstrap-vue'
import { merge } from 'lodash'
import TopSkillChart from '@/components/Dashboard/TopSkillChart'

const localVue = createLocalVue()
localVue.use(BootstrapVue)

function createWrapper (overrideMountingOptions) {
  const defaultMountingOptions = {
    localVue
  }
  const mergedMountingOptions = merge(defaultMountingOptions, overrideMountingOptions)
  return shallowMount(TopSkillChart, mergedMountingOptions)
}

const mockSkillInfo = {
  1: {
    name: 'SkillOne',
    proficiencies: [2],
    willingness: [4],
    proficiencyAverage: 2,
    willingnessAverage: 4
  },
  2: {
    name: 'SkillTwo',
    proficiencies: [1, 5],
    willingness: [0, 5],
    proficiencyAverage: 3,
    willingnessAverage: 2.5
  }
}

describe('TopSkillChart.vue', () => {
  it('Should render chart with text telling information about the top skill', () => {
    const propsData = {
      skillInfo: mockSkillInfo
    }
    const expected = { name: 'SkillTwo', currentKnowledgeAverage: 3, numberOfPeople: 1 }
    const wrapper = createWrapper({ propsData })
    expect(wrapper.vm.topSkillLevel).toEqual(expected)
  })

  it('Should match the template', () => {
    const propsData = {
      skillInfo: mockSkillInfo
    }
    const wrapper = createWrapper({ propsData })
    expect(wrapper.element).toMatchSnapshot()
  })
})
