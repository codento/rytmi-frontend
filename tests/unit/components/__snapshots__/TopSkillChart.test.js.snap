// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopSkillChart.vue Should match the template 1`] = `
<chartcard-stub>
  <template>
    <div
      class="h3 px-2"
    >
      
      Top skill
    
    </div>
  </template>
   
  <bcol-stub
    tag="div"
  >
    <p
      class="lead px-2"
    >
      
      Our top skill is
      
      <strong>
        SkillTwo
      </strong>
      . 
      <br />
      
      There are 
      <strong>
        1
      </strong>
      
      who's skill level is 4 or 5 and the average is
      
      <strong>
        3
      </strong>
    </p>
  </bcol-stub>
</chartcard-stub>
`;
